src/Components/WebAssembly/Server/src/ComponentsWebAssemblyEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Mime;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.StaticFiles;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    /// <summary>
+    /// Extensions for mapping Blazor WebAssembly applications.
+    /// </summary>
+    public static class ComponentsWebAssemblyEndpointRouteBuilderExtensions
+    {
+        /// <summary>
+        /// Maps a Blazor webassembly application to the <paramref name="pathPrefix"/>.
+        /// </summary>
+        /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/>.</param>
+        /// <param name="pathPrefix">The <see cref="PathString"/> that indicates the prefix for the Blazor application.</param>
+        /// <returns>The <see cref="IEndpointConventionBuilder"/></returns>
+        public static IEndpointConventionBuilder MapBlazorWebAssemblyApplication(this IEndpointRouteBuilder endpoints, PathString pathPrefix)
+        {
+            if (endpoints is null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            var webHostEnvironment = endpoints.ServiceProvider.GetRequiredService<IWebHostEnvironment>();
+
+            var options = CreateStaticFilesOptions(webHostEnvironment.WebRootFileProvider);
+            var appBuilder = endpoints.CreateApplicationBuilder();
+
+            appBuilder.Use(async (ctx, next) =>
+            {
+                var endpoint = ctx.GetEndpoint();
+                try
+                {
+                    // Set the endpoint to null so that static files doesn't discard the path.
+                    ctx.SetEndpoint(null);
```

Comment:
I don't know what it means for "static files to discard the path" but I trust this makes sense. For my education, could you briefly let me know why static files would "discard the path"?

Reply:
Yeah, sorry. Static files will not run if it sees the request has an endpoint, so we need to trick it into thinking it doesn't. All MapFallback overloads do this.

Reply:
Makes sense, thanks.